### PR TITLE
Suppress generation of redundant nodes

### PR DESCRIPTION
**Pull request by bifurcation**
Corresponds to https://github.com/mlswg/mls-protocol/pull/587

Depends on #285 and #286

### CODE REVIEW

File: src/key_schedule.cpp [hunk 1]
```diff
@@ -148,23 +148,23 @@ SecretTree::get(LeafIndex sender)
...
-    auto& secret = secrets[curr_node.val];
-    secrets[left.val] =
-      derive_tree_secret(suite, secret, "tree", left, 0, secret_size);
-    secrets[right.val] =
-      derive_tree_secret(suite, secret, "tree", right, 0, secret_size);
+    auto curr_node = dirpath.at(curr);
+    auto left = curr_node.left();
+    auto right = curr_node.right();
+
+    auto& secret = secrets.at(curr_node);
```
Comment (suhasHere) [inline src/key_schedule.cpp:155]: do we want to ensure against this lookup failing ?

**Review by suhasHere — APPROVED** [review]
just one comment .. overall the flow  seems to match the spec update in general

File: src/key_schedule.cpp (hunk 1 above)
Comment (bifurcation) [reply src/key_schedule.cpp:155]: `.at()` throws if the key doesn't exist, so it performs the check for us.

Comment (suhasHere) [reply src/key_schedule.cpp:155]: ah ! that's right .. thanks

**Pull request merged**